### PR TITLE
Exchange currency API

### PR DESCRIPTION
**Pull request by prakashchoudhary07**

### CODE REVIEW

File: models/currency.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+const allCurrencyName = 'allCurrencies'
+
+/**
+ * Fetch the currency exchange rates
+ * @return {Promise<exchangeData>}
+ */
+const fetchExchangeRates = async () => {
+  try {
+    const exchangeSnapshot = await currencyExchangeModel.doc(allCurrencyName).get()
```
Comment (swarajpure) [inline models/currency.js:15]: Don't we pass the document ID in `.doc()`? Are we going to manually set the document ID as `allCurrencies`?

File: models/currency.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        status: false,
+        message: ''
+      }
+
+      const userWalletDoc = await t.get(userWalletRef)
+      const userWalletId = await extractRefDocsId(userWalletDoc)
+      const userWallet = await extractRefDocsData(userWalletDoc)
+      const bankWalletDoc = await t.get(bankWalletRef)
+      const bankWalletId = await extractRefDocsId(bankWalletDoc)
+      const bankWallet = await extractRefDocsData(bankWalletDoc)
```
Comment (swarajpure) [inline models/currency.js:147]: I guess all of these calls will wait for every previous call to complete, slowing down the process (correct me if I'm wrong)

I guess that the `2nd and 3rd` and `5th and 6th` calls can be solved using `Promise.all` . What do you say? I don't have that good understanding of promises, so let me know if I'm saying anything wrong 😄

File: routes/exchange.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+const router = express.Router()
+const exchangeController = require('../controllers/exchangeController')
+const authenticate = require('../middlewares/authenticate')
+const exchangeVaidatore = require('../middlewares/validators/exchange')
+
+/**
+ * @swagger
+ * /exchange/rates:
+ *   get:
+ *     summary: Get exhcange rate
```
Comment (swarajpure) [inline routes/exchange.js:11]: NIT: `exchange`

File: routes/exchange.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ *             schema:
+ *               $ref: '#/components/schemas/errors/badImplementation'
+ */
+router.get('/rates', exchangeController.getExchangeRate)
+
+/**
+ * @swagger
+ * /exchange/rates:
+ *   get:
+ *     summary: post exhcange rate
```
Comment (swarajpure) [inline routes/exchange.js:34]: NIT: `exchange`

**Review by swarajpure — CHANGES_REQUESTED** [review]
Great job there! The OG FTL ⚡ 

Having some queries. Please check.

File: controllers/exchangeController.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    // This id temperory hack, to make ankuhs set the exchange rate
+    if (username === 'ankush') {
+      const currencyData = { ...req.body, timestamp: new Date() }
+      createCurrencyRates(currencyData)
+      return res.json({
+        message: 'currency created successfully!',
+        banks: currencyData
+      })
+    }
+    return res.boom.unauthorized('You are not authorized to use this, only ankush can XD')
```
[the remark this answers is not in the document]
Comment (harshith-venkatesh) [reply controllers/exchangeController.js:37]: I loved this too

File: models/currency.js [hunk 6]
```diff
@@ -0,0 +1,189 @@
+const firestore = require('../utils/firestore')
+const walletModel = firestore.collection('wallets')
+const bankModel = firestore.collection('banks')
+const currencyExchangeModel = firestore.collection('currencyExchange')
+// const transactionModel = firestore.collection('transactions')
```
[the remark this answers is not in the document]
Comment (harshith-venkatesh) [reply models/currency.js:5]: This comes in //todo right prakash?

File: models/currency.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+/**
+ * Extracts the data form the first doc of firestore response
+ * @return {Promise<object>}
+ */
+const extractRefDocsData = (doc) => {
+  return doc.docs[0] && doc.docs[0].data()
+}
+
+/**
+ * Gets the id of frestore documenr
```
Comment (harshith-venkatesh) [inline models/currency.js:98]: NIT : firestore

**Review by harshith-venkatesh — APPROVED** [review]

File: models/currency.js (hunk 2 above)
Comment (prakashchoudhary07) [reply models/currency.js:147]: All read and writes are taking place in transaction

File: models/currency.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply models/currency.js:5]: Yes, it will be build and pr will be sent for it ASAP

**Review by harshith-venkatesh — DISMISSED** [review]

File: models/currency.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+      }
+
+      const userWalletDoc = await t.get(userWalletRef)
+      const userWalletId = await extractRefDocsId(userWalletDoc)
+      const userWallet = await extractRefDocsData(userWalletDoc)
+      const bankWalletDoc = await t.get(bankWalletRef)
+      const bankWalletId = await extractRefDocsId(bankWalletDoc)
+      const bankWallet = await extractRefDocsData(bankWalletDoc)
+
+      if (bankWallet && bankWallet.isActive) {
```
Comment (Kratika0907) [inline models/currency.js:149]: Can we pull these into separate functions? 
Also, can you point to the data model and API contracts for same

File: models/currency.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kratika0907) [reply models/currency.js:5]: Has this been updated? We need to update transactions for users and banks also.
For bank: @ankushdharkar Do we want to maintain the record for the same.

**Review by Kratika0907 — CHANGES_REQUESTED** [review]

File: models/currency.js [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+  bankCurrency[exchangeData.src] = (bankCurrency[exchangeData.src] || 0) + parseInt(exchangeData.quantity)
+  bankCurrency[exchangeData.target] -= totalTargetCurrencyRequest
+  return { ...bankCurrency }
+}
+
+/**
+ * Update user currecy after excbange
+ * @return {Promise<exchangeData>}
+ */
+const updateUserWalletAfterExchange = (userCurrency, exchangeData, totalTargetCurrencyRequest) => {
```
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply models/currency.js:119]: This is done using transaction, so the existing function won't fit in properly :sweat_smile:

File: models/currency.js [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+ * Transaction for currency exchange
+ * @return {Promise<exchangeData>}
+ */
+const exchangeTransaction = async (userId, exchangeData) => {
+  try {
+    const { bankId } = exchangeData
+    // db models refs
+    const userWalletRef = await walletModel.where('userId', '==', userId).limit(1)
+    const bankWalletRef = await bankModel.where('bankId', '==', bankId).limit(1)
+    const exchangeRateRef = await currencyExchangeModel.doc(allCurrencyName)
```
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply models/currency.js:135]: It's done in transaction so it shouldn't matter :sweat_smile:

File: models/currency.js (hunk 8 above)
Comment (prakashchoudhary07) [reply models/currency.js:149]: https://github.com/Real-Dev-Squad/website-api-contracts/pull/46

File: controllers/exchangeController.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * Create exchange rates
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const createExchangeRate = (req, res) => {
+  try {
+    const { username } = req.userData
+    // This id temporary hack, to make ankush set the exchange rate
+    if (username === 'ankush') {
```
Comment (ankushdharkar) [inline controllers/exchangeController.js:29]: Use the middleware here please

File: controllers/exchangeController.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+      const currencyData = { ...req.body, timestamp: new Date() }
```
Comment (ankushdharkar) [inline controllers/exchangeController.js:30]: Let's change this to milliseconds since epoch: `Date.now()`

File: models/currency.js [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+ * @return {Promise<exchangeData>}
+ */
+const createCurrencyRates = async (currencyData) => {
+  try {
+    await currencyExchangeModel.add(currencyData)
+    await currencyExchangeModel.doc(allCurrencyName).set({
+      [currencyData.src]: {
+        [currencyData.target]: currencyData.quantity
+      },
+      lastUpdate: new Date()
```
Comment (ankushdharkar) [inline models/currency.js:57]: `Date.now()`

File: models/currency.js [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+  return doc.docs[0] && doc.docs[0].data()
+}
+
+/**
+ * Gets the id of firestore document
+ * @return {Promise<object>}
+ */
+const extractRefDocsId = (doc) => {
+  return doc.docs[0] && doc.docs[0].id
+}
```
Comment (ankushdharkar) [inline models/currency.js:103]: Move to firestore utils?

File: models/currency.js [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+
+/**
+ * Update bank currecy after excbange
+ * @return {Promise<exchangeData>}
+ */
+const updateBankWalletAfterExchange = (bankCurrency, exchangeData, totalTargetCurrencyRequest) => {
+  bankCurrency[exchangeData.src] = (bankCurrency[exchangeData.src] || 0) + parseInt(exchangeData.quantity)
+  bankCurrency[exchangeData.target] -= totalTargetCurrencyRequest
```
Comment (ankushdharkar) [inline models/currency.js:111]: Should value be going from source to target?

File: models/currency.js [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+ * @return {Promise<exchangeData>}
+ */
+const exchangeTransaction = async (userId, exchangeData) => {
+  try {
+    const { bankId } = exchangeData
+    // db models refs
+    const userWalletRef = await walletModel.where('userId', '==', userId).limit(1)
+    const bankWalletRef = await bankModel.where('bankId', '==', bankId).limit(1)
+    const exchangeRateRef = await currencyExchangeModel.doc(allCurrencyName)
+    const exchangeResponse = await firestore.runTransaction(async t => {
```
Comment (ankushdharkar) [inline models/currency.js:136]: Nice use of firestore transactions!

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: controllers/exchangeController.js [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+ * Exchange currency in between user and bank
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const convertCurrency = async (req, res) => {
+  try {
+    const exchangeData = req.body
+    const { id: userId } = req.userData
+    const exchangeTransactionStatus = await exchangeTransaction(userId, exchangeData)
+    return res.status(200).json({
```
Comment (swarajpure) [inline controllers/exchangeController.js:89]: `status(200)` not required here

File: controllers/exchangeController.js [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * Create exchange rates
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const createExchangeRate = (req, res) => {
+  try {
+    const currencyData = { ...req.body, timestamp: Date.now() }
+    createCurrencyRates(currencyData)
```
Comment (swarajpure) [inline controllers/exchangeController.js:28]: No `await` here? 🤔

File: docs/swaggerDefinition.js [hunk 19]
```diff
@@ -579,6 +579,34 @@ const swaggerOptions = {
...
+            isActive: {
+              type: 'boolean'
+            },
+            quantity: {
+              type: 'number'
+            },
+            currency: {
+              type: 'object',
+              properties: {
+                denaru: {
```
Comment (swarajpure) [inline docs/swaggerDefinition.js:600]: `denaru`? 🤔 
Can we have it from constants?

File: models/currency.js [hunk 20]
```diff
@@ -0,0 +1,188 @@
+const firestore = require('../utils/firestore')
+const walletModel = firestore.collection('wallets')
+const bankModel = firestore.collection('banks')
+const currencyExchangeModel = firestore.collection('currencyExchange')
+const transactionModel = firestore.collection('transactions')
+const { extractReferenceDocumentData, extractReferenceDocumentId } = require('../utils/firestore-helper')
+
+const allCurrencyName = 'allCurrencies'
```
Comment (swarajpure) [inline models/currency.js:8]: Can we have this in constants file? If someone else needs it in future in some other place?

File: models/currency.js [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+ * @return {Promise<exchangeData>}
+ */
+const getAllBanks = async () => {
+  try {
+    const bankSnapshot = await bankModel.get()
+    const bankData = []
+    bankSnapshot.forEach(bank => {
+      const data = bank.data()
+      if (data.isActive) {
+        bankData.push({ bankId: data.bankId, bankName: data.bankName })
```
Comment (swarajpure) [inline models/currency.js:80]: Could we destructure `bankId` and `bankName` and use? Wouldn't it increase the readability? What do you think?

File: models/currency.js [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+  } catch (err) {
+    logger.error('Error while getting bank detais', err)
+    throw err
+  }
+}
+
+/**
+ * Rounds number to 2 precision after decimal point
+ * @return {Promise<object>}
+ */
```
Comment (swarajpure) [inline models/currency.js:93]: Are you sure it returns a promise? 🤔 
image

Source: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/round

File: models/currency.js [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+        status: false,
+        message: ''
+      }
+
+      const userWalletDoc = await t.get(userWalletRef)
+      const userWalletId = await extractReferenceDocumentId(userWalletDoc)
+      const userWallet = await extractReferenceDocumentData(userWalletDoc)
+      const bankWalletDoc = await t.get(bankWalletRef)
+      const bankWalletId = await extractReferenceDocumentId(bankWalletDoc)
+      const bankWallet = await extractReferenceDocumentData(bankWalletDoc)
```
Comment (swarajpure) [inline models/currency.js:140]: Feel free to correct me here

Initally promise corresponding to `userWalletDoc` will resolve, then `userWalletId` then `userWallet` and then `bankWalletDoc` would get an opportunity to start executing.
`bankWalletDoc` has nothing to do with `userWalletDoc` or `userWalletId` or `userWallet`, right?
Aren't we then unnecessarily blocking the execution here?

File: routes/exchange.js [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+ *         content:
+ *           application/json:
+ *             schema:
+ *               $ref: '#/components/schemas/errors/badImplementation'
+ */
+router.get('/banks', exchangeController.getAllBanksName)
+
+/**
+ * @swagger
+ * /exchange/{bankId}::
```
Comment (swarajpure) [inline routes/exchange.js:85]: NIT: `/exchange/:bankId`

File: routes/exchange.js [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+ *         content:
+ *           application/json:
+ *             schema:
+ *               $ref: '#/components/schemas/errors/badImplementation'
+ */
+router.get('/:bankId', exchangeController.getCurrencyAvailabeInBank)
+
+/**
+ * @swagger
+ * /exchange/{bankId}::
```
Comment (swarajpure) [inline routes/exchange.js:114]: NIT: `/exchange/:bankId`

**Review by swarajpure — CHANGES_REQUESTED** [review]
Have some doubts and some suggestions. Please check.

Thank you! 😄

File: docs/swaggerDefinition.js (hunk 19 above)
Comment (prakashchoudhary07) [reply docs/swaggerDefinition.js]: Forgot :sweat:

File: models/currency.js (hunk 22 above)
Comment (prakashchoudhary07) [reply models/currency.js]: Copy pasted the jsdoc, forgot to change it :sweat_smile:

File: controllers/exchangeController.js (hunk 18 above)
Comment (prakashchoudhary07) [reply controllers/exchangeController.js]: Forgot :sweat:

File: constants/wallets.js [hunk 26]
```diff
@@ -3,5 +3,6 @@ module.exports = {
   NEELAM: 'neelam',
   INITIAL_WALLET: {
     dinero: 1000
-  }
+  },
+  SUPPORTED_CURRENCIES: ['dinero', 'neelam']
```
Comment (swarajpure) [inline constants/wallets.js:7]: This file needs a bit of restructuring I guess so we can reuse the DINERO and NEELAM described above

`CURRENCY_TYPES` could contain 
```JS
{ 
  DINERO: 'dinero',
  NEELAM: 'neelam'
}
```

Then we can use the values of the object `CURRENCY_TYPES` in `SUPPORTED_CURRENCIES`

File: controllers/exchangeController.js [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+ * Create exchange rates
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const createExchangeRate = async (req, res) => {
+  try {
+    const currencyData = { ...req.body, timestamp: Date.now() }
+    await createCurrencyRates(currencyData)
+    return res.json({
+      message: 'currency created successfully!',
```
Comment (swarajpure) [inline controllers/exchangeController.js:30]: NIT: capitalize first letter of the message?

File: controllers/exchangeController.js [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const getCurrencyAvailabeInBank = async (req, res) => {
+  try {
+    const bankId = req.params.bankId
+    const currencyData = await fetchCurrencyAvailablity(bankId)
+    if (currencyData) {
+      return res.json({
+        message: 'currency returned successfully!',
```
Comment (swarajpure) [inline controllers/exchangeController.js:50]: NIT: capitalize first letter of the message?

File: controllers/exchangeController.js [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+/**
+ * Get the bank's currency details
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const getCurrencyAvailabeInBank = async (req, res) => {
+  try {
+    const bankId = req.params.bankId
```
Comment (swarajpure) [inline controllers/exchangeController.js:46]: destructuring?

File: models/currency.js [hunk 30]
```diff
@@ -0,0 +1,224 @@
...
+
+/**
+ * Rounds number to 2 precision after decimal point
+ * @param num { Number }
+ * @return {Number}
+ */
+const roundToTwoDigits = (num) => {
+  return Math.round(num * 100) / 100
+}
+
```
Comment (swarajpure) [inline models/currency.js:100]: Shouldn't this go in `utils`? 🤔 
Maybe in `utils/math`?

File: routes/exchange.js [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+ *             schema:
+ *               $ref: '#/components/schemas/errors/unAuthorized'
+ *       500:
+ *         description: badImplementation
+ *         content:
+ *           application/json:
+ *             schema:
+ *               $ref: '#/components/schemas/errors/badImplementation'
+ */
+router.post('/rates', authenticate, authorizeUser('superUser'), postCurrencyRates, exchangeController.createExchangeRate)
```
Comment (swarajpure) [inline routes/exchange.js:58]: `superUser` can come from constants?
Implemented in this PR: https://github.com/Real-Dev-Squad/website-backend/pull/357
You can copy paste maybe?

**Review by swarajpure — CHANGES_REQUESTED** [review]
Some suggestions for constants, destructuring and some NIT comments.
Please check.

File: models/exchange.js [hunk 32]
```diff
@@ -0,0 +1,216 @@
...
+ * @return {Promise<exchangeData>}
+ */
+const exchangeTransaction = async (userId, exchangeData) => {
+  try {
+    const { bankId } = exchangeData
+    // db models refs
+    const userWalletRef = walletModel.where('userId', '==', userId).limit(1)
+    const bankWalletRef = bankModel.where('bankId', '==', bankId).limit(1)
+    const exchangeRateRef = currencyExchangeModel.doc(allCurrencyName)
+    const exchangeResponse = await firestore.runTransaction(async t => {
```
Comment (sumitd94) [inline models/exchange.js:149]: `t` is pretty vague, can we please use like `transac` or `tran`

File: utils/firestore-helper.js [hunk 33]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Extracts the data form the first doc of firestore response
+ * @return {Promise<object>}
+ */
+const extractReferenceDocumentData = (doc) => {
+  return doc.docs[0] && doc.docs[0].data()
+}
```
Comment (sumitd94) [inline utils/firestore-helper.js:7]: NIT: can convert it into a one-liner

```javascript
const extractReferenceDocumentData = (doc) => doc.docs[0] && doc.docs[0].data()
```

File: utils/firestore-helper.js [hunk 34]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * Gets the id of firestore document
+ * @return {Promise<object>}
+ */
+const extractReferenceDocumentId = (doc) => {
+  return doc.docs[0] && doc.docs[0].id
+}
```
Comment (sumitd94) [inline utils/firestore-helper.js:15]: RE:

File: utils/math-helpers.js [hunk 35]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Rounds number to 2 precision after decimal point
+ * @param num { Number }
+ * @return {Number}
+ */
+const roundToTwoDigits = (num) => {
+  return Math.round(num * 100) / 100
+}
```
Comment (sumitd94) [inline utils/math-helpers.js:8]: NIT: oneliner

**Review by sumitd94 — CHANGES_REQUESTED** [review]
Few comments added and small changes requested, please

File: models/exchange.js [hunk 36]
```diff
@@ -0,0 +1,216 @@
...
+ * Fetch the currency exchange rates
+ * @return {Promise<exchangeData>}
+ */
+const fetchExchangeRates = async () => {
+  try {
+    const exchangeSnapshot = await currencyExchangeModel.doc(allCurrencyName).get()
+    if (exchangeSnapshot.exists) {
+      return exchangeSnapshot.data()
+    }
+    return undefined
```
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply models/exchange.js:21]: refactored it

**Pull request closed without merging**